### ISSUE TITLE
[Feature Request] Provide Migrations API

### ISSUE THREAD

**Issue by tylergets** [zenstackhq/zenstack#1258]
**Is your feature request related to a problem? Please describe.**
Prisma itself does not offer programmatic access to its migration library, originally reported here: https://github.com/prisma/prisma/issues/4703
 
**Describe the solution you'd like**
It would be great if Zenstack provided a way to run migrations programmaticly. 

**Describe alternatives you've considered**
Running the prisma command via a shell exec

**Additional context**
The main purpose of this for me is to allow applications to bootstrap itself without needing to run the prisma db push command or ship a preseeded DB

**Comment by ymc9** [comment]
Hi @tylergets ,

ZenStack doesn't integrate with Prisma migrate in any way today. If it does that I think it should do it in a way that's integrated with other core features (access control etc.).

I'm keeping this issue in the backlog for now to collect more feedback.

**Comment by andrictham** [comment]
I think having a good workflow around migrations, especially data migrations, are going to be necessary in order to use generated CRUD hooks in a production app.

As they stand now, version skew could cost queries to fail, if a client uses an old version of generated hooks tied to an old schema. 

The best practice to handle version skew of APIs is to version your APIs, so you can maintain the same shape of response for older clients. 

How do you do that with the auto-generated RPC-style APIs?

I know that I can still manually create my own API routes using tRPC for instance, and calling the enhance() function.

But for developer productivity, it would be nice if this versioning could be handled for me, since ZenStack knows about my schema and it knows how it changes over time. 

Currently [data migrations](https://www.prisma.io/docs/orm/prisma-migrate/workflows/data-migration) are a missing piece of the Prisma workflow. But Prisma doesn’t handle CRUD API generation. We’re expected to implement our own.

But with ZenStack, let's say we follow the “expand and contract”recommended by Prisma to migrate our data, and we drop the column in the “contract” step, wouldn’t  old clients break because the frontend hooks mirror the Prisma API which mirrors the database columns?

The way the frontend query code is coupled to the database schema is not ideal in this scenario, because it makes the API resistant to change.

**Comment by ymc9** [comment]
Hey @andrictham , thanks for the wait! I think I don't fully understand the versioning scenario yet.

Here's my confusion. In the "expand and contract" process, when you take the "contract" step and finally remove the old column, there's supposed to be no one using that column anymore, right? In the context of ZenStack, the backend enhanced prisma should have been updated in previous deployments to cut over from reading/writing the old columns to the new ones. The frontend hooks should have been regenerated and redeployed, too.

Is this specifically about the browser windows that haven't been refreshed to load the new frontend bundle yet? Or is it about a general mechanism to run multiple API versions that serve different versions of schemas concurrently? So that you can have the client to keep running different versions of hooks for an extended period of time?

**Comment by andrictham** [comment]
@ymc9 

> Is this specifically about the browser windows that haven't been refreshed to load the new frontend bundle yet? 

Yes, mostly, that would be my biggest concern for a live production app, as there is no good way to guarantee users will not be using an older version, leading to version skew between client and server, unless we do a version check and force clients to refresh.

This is enough of a problem that [Vercel specifically has a feature to handle this](https://vercel.com/blog/version-skew-protection). 

Vercel’s version skew wouldn’t work here, though, because the older backend code would break too since it’s tightly coupled to the database schema.

This _could_ become a non-issue if I do a version check and/or wait long enough for clients to phase out of using older client code. 

But that strategy would only work if we forgo any sort of offline capability with optimistic mutations, which Tanstack Query supports.

> Or is it about a general mechanism to run multiple API versions that serve different versions of schemas concurrently? So that you can have the client to keep running different versions of hooks for an extended period of time?

I don’t know if there's an easier way to solve version skew, but if there’s an easier way, I’m all for it.

There seems to be two separate issues here:
- Version skew between client and server (our RPC hooks)
- Version skew between server and database (our Prisma client)

Since ZenStack uses a single schema to generate both, it also has the potential ability to resolve both types of version skew. 

And at the same time, the ability to serve multiple backward-compatible API versions and handle deprecated or renamed fields gracefully would also be a win for developer productivity.

There are also migration issues where a schema change would require existing data or older requests to be migrated to the new schema, which Prisma doesn’t handle today. 

Since ZenStack is schema-first and already a DSL for writing custom rules (now largely validation and access control rules), I can imagine it being able to also support data migration rules. 

This could help simplify simple schema changes like renaming or combining/splitting fields, or complex tasks like relaxing or imposing new constraints on the data or changing data types.

Off the top of my head, this would be rules like how to:
- Backfill newly-required columns that used to be optional with a default value
- Reroute requests to a column or model that was renamed, using aliases to handle requests coming in that still refer to the column or model by its old name
- Gracefully handle cases where a column in a previous schema had different validation rules or data type, and how to reformat it to be compatible with the new rules/type
- Gracefully handle cases where one or more columns are merged into a single column, by writing rules to transform data, like a computed value (say `firstName + lastName`)
- Gracefully handle cases where a single column is split into multiple columns, by writing rules to transform data (splitting by substring, for instance)

And for version skew between client and server in RPC code, perhaps older clients automatically send a version which signals what schema version they are querying or mutating, which would trigger the relevant data migration rules to run.
